### PR TITLE
Improve the XML BBEs

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

### CODE REVIEW

File: examples/xml-namespaces/xml_namespaces.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
 # XML namespaces
 
-The goal is to support namespaces without adding complexity if you don’t use them. The 
-`ns:x` qualified name in XML is expanded into `{url}x` where `url` is the namespace name bound to `ns`. The XML namespace
-declarations are kept as attributes using the standard binding of [xmlns](http://www.w3.org/2000/xmlns/).
+The `ns:x` qualified name in XML is expanded into `{url}x` where `url` is the namespace name bound to `ns`. The XML namespace declarations are kept as attributes using the standard binding of [xmlns](http://www.w3.org/2000/xmlns/).
+
+XML namespaces are supported in such a way that they do not add complexity if you do not use them. 
 
 ::: code xml_namespaces.bal :::
```
Comment (gimantha) [inline examples/xml-namespaces/xml_namespaces.md:8]: Don't we need to say that executing above code, give the below result? I have seen this in new format

Comment (MaryamZi) [reply examples/xml-namespaces/xml_namespaces.md:8]: I believe that's only for the first one or two BBEs. We had it in the original BBEs too, but decided to remove because it was kinda repetitive and not adding value IIRC.

File: examples/xml-operations/xml_operations.bal [hunk 2]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
...
     io:println(name is ());
 
     // Mutate an element using `e.setChildren(x)`.
     x2.setChildren(xml `<language>French</language>`);
 
+    // The `xml` value assigned to `x2` will now include `<language>French</language>` as a 
+    // child element.
     io:println(x2);
+
+    // Since the value assigned to `x3` used `x2` in the concatenation, the change will also
```
Comment (KRVPerera) [inline examples/xml-operations/xml_operations.bal:58]:
```suggestion
    // Since the value assigned to `x3` used `x2` in the concatenation, `x2.setChildren` will also change `x3`
```

File: examples/xml-operations/xml_operations.bal [hunk 3]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
...
 
     // Mutate an element using `e.setChildren(x)`.
     x2.setChildren(xml `<language>French</language>`);
 
+    // The `xml` value assigned to `x2` will now include `<language>French</language>` as a 
+    // child element.
     io:println(x2);
+
+    // Since the value assigned to `x3` used `x2` in the concatenation, the change will also 
+    // be reflected in it.
```
Comment (KRVPerera) [inline examples/xml-operations/xml_operations.bal:59]: Related to previous one
```suggestion
```

File: examples/xml-subtyping/xml_subtyping.bal [hunk 4]
```diff
@@ -1,29 +1,34 @@
...
     string world = "World";
+
+    // The return type of `stringToXml` is `xml:Text`, which indicates that it will
+    // return an XML text item.
     xml:Text c = stringToXml(hello + " " + world);
     io:println(c);
 
     xml:Element otherElement = xml `<q>World</q>`;
 
+    // Concatenating multiple items results in a sequence of items (`xml<T>`).
```
Comment (KRVPerera) [inline examples/xml-subtyping/xml_subtyping.bal:26]: Do we need this comment? I don't think user can understand the comment looking at the code.

File: examples/xml-subtyping/xml_subtyping.bal [hunk 5]
```diff
@@ -1,29 +1,34 @@
...
     xml d = element + otherElement;
+
     xml e = xml `<p>hello</p>World`;
-    // An `xml` value belongs to the type `xml<T>` if each of its members belong 
-    // to `T`.
+
+    // An `xml` value belongs to the type `xml<T>` if each of its members belong to `T`.
```
Comment (KRVPerera) [inline examples/xml-subtyping/xml_subtyping.bal:31]:
```suggestion
    // An `xml` value belongs to the type `xml<T>` if each member belongs to `T`.
```

File: examples/xml-operations/xml_operations.bal [hunk 6]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
...
-    // `x.id` accesses required attribute named `id`:
-    // result is `error` if there is no such attribute
-    // or if `x` is not a singleton.
+    // `x.id` accesses a required attribute named `id`: result is `error` if there is no such 
+    // attribute or if `x` is not a singleton.
     xml x5 = xml `<para id="greeting">Hello</para>`;
     string id = check x5.id;
 
+    // Since an attribute named `id` exists in the `xml` value on which required attribute access 
+    // is done the result of the access will be the value of the attribute (`"greeting"`).
```
[the remark this answers is not in the document]
Comment (MaryamZi) [reply examples/xml-operations/xml_operations.bal:39]: Any particular reason to change to "attribute's value"?

File: examples/xml-operations/xml_operations.bal [hunk 7]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
     // `x[i]` gives i-th item (empty sequence if none).
     io:println(x3[0]);
 
-    // `x.id` accesses required attribute named `id`:
-    // result is `error` if there is no such attribute
-    // or if `x` is not a singleton.
+    // `x.id` accesses a required attribute named `id`: result is `error` if there is no such
```
Comment (MaryamZi) [inline examples/xml-operations/xml_operations.bal:33]:
```suggestion
    // `x.id` accesses a required attribute named `id`: the result is `error` if there is no such 
```

File: examples/xml-subtyping/xml_subtyping.md [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
-An `xml` value belongs to `xml:Text` if it consists of a text item or is 
-empty.
+An `xml` value belongs to `xml:Element` if it consists of just an element item. Similarly an `xml` value belongs to `xml:Comment` or `xml:ProcessingInstruction` if it consists of just a comment item or a processing instruction item respectively.
 
-An `xml` value belongs to the type `xml<T>` if each of its members belong
-to `T`. Functions in the `lang.xml` module use this to provide safe and convenient typing. For example, `x.elements()` returns element items in `x` as type `xml<xml:Element>`. The `e.getName()` and `e.setName()` are defined when
-`e` has the `xml:Element` type.
+An `xml` value belongs to `xml:Text` if it consists of only a text item or is empty.
+
+An `xml` value belongs to the type `xml<T>` if each of its members belong to `T`.
```
Comment (KRVPerera) [inline examples/xml-subtyping/xml_subtyping.md:7]:
```suggestion
An `xml` value belongs to the type `xml<T>` if each member belongs to `T`. 
```

File: examples/xml-operations/xml_operations.bal (hunk 2 above)
Comment (MaryamZi) [reply examples/xml-operations/xml_operations.bal:58]: Used "reflected in it" because it is basically the same value that changes. And `x3` having a reference to the same value is why the change is reflected in `x3`.

File: examples/xml-namespaces/xml_namespaces.bal [hunk 9]
```diff
@@ -1,8 +1,9 @@
 import ballerina/io;
 
-xml:Element e = xml `<p:e xmlns:p="http://example.com/"/>`;
-
 public function main() {
+    // The `p:e` qualified name, where `http://example.com/` is the namespace name bound to `p`,
```
Comment (praneesha) [inline examples/xml-namespaces/xml_namespaces.bal:4]:
```suggestion
    // The `p:e` qualified name of which `http://example.com/` is the namespace name bound to `p`,
```

File: examples/xml-namespaces/xml_namespaces.bal [hunk 10]
```diff
@@ -1,8 +1,9 @@
...
+    // is expanded into `{http://example.com/}e`.
```
Comment (praneesha) [inline examples/xml-namespaces/xml_namespaces.bal:5]:
```suggestion
    // which is expanded into `{http://example.com/}e`.
```

File: examples/xml-operations/xml_operations.bal (hunk 6 above)
[the remark this answers is not in the document]
Comment (KRVPerera) [reply examples/xml-operations/xml_operations.bal:39]: I think not. I thought it is simple but looks like it is same as before.

File: examples/xml-namespaces/xml_namespaces.md [hunk 11]
```diff
@@ -1,9 +1,9 @@
 # XML namespaces
 
-The goal is to support namespaces without adding complexity if you don’t use them. The 
-`ns:x` qualified name in XML is expanded into `{url}x` where `url` is the namespace name bound to `ns`. The XML namespace
-declarations are kept as attributes using the standard binding of [xmlns](http://www.w3.org/2000/xmlns/).
+The `ns:x` qualified name in XML is expanded into `{url}x` where `url` is the namespace name bound to `ns`. The XML namespace declarations are kept as attributes using the standard binding of [xmlns](http://www.w3.org/2000/xmlns/).
```
Comment (praneesha) [inline examples/xml-namespaces/xml_namespaces.md:3]:
```suggestion
The `ns:x` qualified name in XML is expanded into `{url}x` in which the `url` is the namespace name bound to `ns`. The XML namespace declarations are kept as attributes using the standard binding of [`xmlns`](http://www.w3.org/2000/xmlns/).
```

File: examples/xml-operations/xml_operations.bal [hunk 12]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
     // `x[i]` gives i-th item (empty sequence if none).
```
Comment (praneesha) [inline examples/xml-operations/xml_operations.bal:30]:
```suggestion
    // `x[i]` gives the `i-th` item (empty sequence if none).
```

File: examples/xml-subtyping/xml_subtyping.bal (hunk 5 above)
Comment (KRVPerera) [reply examples/xml-subtyping/xml_subtyping.bal:31]: I believe this makes the sentence more simple

File: examples/xml-subtyping/xml_subtyping.md (hunk 8 above)
Comment (KRVPerera) [reply examples/xml-subtyping/xml_subtyping.md:7]: I believe this makes the sentence more simple

File: examples/xml-operations/xml_operations.bal [hunk 13]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
...
-    // `x.id` accesses required attribute named `id`:
-    // result is `error` if there is no such attribute
-    // or if `x` is not a singleton.
+    // `x.id` accesses a required attribute named `id`: the result is `error` if there is no such 
+    // attribute or if `x` is not a singleton.
     xml x5 = xml `<para id="greeting">Hello</para>`;
     string id = check x5.id;
 
+    // Since an attribute named `id` exists in the `xml` value on which required attribute access 
+    // is done the result of the access will be the value of the attribute (`"greeting"`).
```
Comment (praneesha) [inline examples/xml-operations/xml_operations.bal:39]:
```suggestion
    // is done, the result of the access will be the value of the attribute (`"greeting"`).
```

File: examples/xml-operations/xml_operations.bal [hunk 14]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
...
     io:println(id);
 
-    // `x?.id` accesses optional attribute named `id`:
-    // result is `()` if there is no such attribute.
+    // `x?.id` accesses an optional attribute named `id`: the result is `()` if there is no such 
+    // attribute.
     string? name = check x5?.name;
 
+    // Since an attribute named `name` does not exist in the `xml` value on which optional 
+    // attribute access is done the result of the access is nil. Therefore the following
```
Comment (praneesha) [inline examples/xml-operations/xml_operations.bal:47]:
```suggestion
    // attribute access is done, the result of the access is nil. Therefore, the following
```

File: examples/xml-operations/xml_operations.bal (hunk 6 above)
[the remark this answers is not in the document]
Comment (KRVPerera) [reply examples/xml-operations/xml_operations.bal:39]: Shall I change the edit?

File: examples/xml-operations/xml_operations.md [hunk 15]
```diff
@@ -1,13 +1,14 @@
...
-- `x.id` accesses required attribute named `id`. The result will be an error if there is no such attribute
-  or if x is not a singleton.
-- `x?.id` accesses optional attribute named `id`. The result will be `()` if there is no such attribute.
-- [XML language library](https://lib.ballerina.io/ballerina/lang.xml/0.0.0) provides many other ways to manipulate XML using `lang.xml` module. For an example, using the langlib function `e.setChildren(x)` an XML element can be mutated.
+Numerous operations including those listed below can be performed on `xml` values.
+
+- `+` does concatenation
+- `==` does a deep equality check
+- `foreach` iterates over each item
+- `x[i]` gives the i-th item (empty sequence if none)
```
Comment (praneesha) [inline examples/xml-operations/xml_operations.md:8]:
```suggestion
- `x[i]` gives the `i-th` item (empty sequence if none)
```

File: examples/xml-operations/xml_operations.md [hunk 16]
```diff
@@ -1,13 +1,14 @@
...
+- `x.id` accesses a required attribute named `id`. The result will be an error if there is no such attribute or if x is not a singleton.
```
Comment (praneesha) [inline examples/xml-operations/xml_operations.md:9]:
```suggestion
- `x.id` accesses a required attribute named `id`. The result will be an error if there is no such attribute or if `x` is not a singleton.
```

File: examples/xml-subtyping/xml_subtyping.bal [hunk 17]
```diff
@@ -1,29 +1,34 @@
 import ballerina/io;
 
-// An `xml` value belongs to `xml:Element` if it consists of just an element
-// item. 
+// An `xml` value belongs to `xml:Element` if it consists of just an element item. 
 xml:Element element = xml `<p>Hello</p>`;
 
-// Similarly for `xml:Comment` and `xml:ProcessingInstruction`.
+// Similarly a value belongs to `xml:Comment` or `xml:ProcessingInstruction` if it
```
Comment (praneesha) [inline examples/xml-subtyping/xml_subtyping.bal:6]:
```suggestion
// Similarly, a value belongs to `xml:Comment` or `xml:ProcessingInstruction` if it 
```

File: examples/xml-operations/xml_operations.bal (hunk 2 above)
Comment (KRVPerera) [reply examples/xml-operations/xml_operations.bal:58]: let me try again

File: examples/xml-subtyping/xml_subtyping.bal (hunk 5 above)
Comment (praneesha) [inline examples/xml-subtyping/xml_subtyping.bal:31]:
```suggestion
    // An `xml` value belongs to the  `xml<T>` type if each of its members belongs to `T`.
```

File: examples/xml-subtyping/xml_subtyping.md [hunk 18]
```diff
@@ -1,14 +1,14 @@
 # XML subtyping
 
-An `xml` value belongs to `xml:Element` if it consists of just an element
-item. The same rule applies for `xml:Comment` and `xml:ProcessingInstruction`.
-An `xml` value belongs to `xml:Text` if it consists of a text item or is 
-empty.
+An `xml` value belongs to `xml:Element` if it consists of just an element item. Similarly an `xml` value belongs to `xml:Comment` or `xml:ProcessingInstruction` if it consists of just a comment item or a processing instruction item respectively.
```
Comment (praneesha) [inline examples/xml-subtyping/xml_subtyping.md:3]:
```suggestion
An `xml` value belongs to `xml:Element` if it consists of just an element item. Similarly, an `xml` value belongs to `xml:Comment` or `xml:ProcessingInstruction` if it consists of just a comment item or a processing instruction item respectively.
```

File: examples/xml-operations/xml_operations.bal [hunk 19]
```diff
@@ -30,23 +30,32 @@ public function main() returns error? {
...
 
     // Mutate an element using `e.setChildren(x)`.
     x2.setChildren(xml `<language>French</language>`);
 
+    // The `xml` value assigned to `x2` will now include `<language>French</language>` as a 
+    // child element.
     io:println(x2);
+
+    // Since the value assigned to `x3` used `x2` in the concatenation, the change will also 
+    // be reflected in it.
```
Comment (KRVPerera) [inline examples/xml-operations/xml_operations.bal:59]:
```suggestion
    // be reflected in `x3`.
```

File: examples/xml-subtyping/xml_subtyping.md (hunk 8 above)
Comment (praneesha) [inline examples/xml-subtyping/xml_subtyping.md:7]:
```suggestion
An `xml` value belongs to the type `xml<T>` if each of its members belongs to `T`. 
```

File: examples/xmlns-declarations/xmlns_declarations.bal [hunk 20]
```diff
@@ -1,17 +1,25 @@
...
 
-xmlns "http://example.com" as ex;
-
 public function main() {
+    // The prefixes to which namespaces declarations are bound to can be used in XML templates.
     xml x = xml `<eg:doc>Hello</eg:doc>`;
 
-    boolean b = (x === x.<ex:doc>);
+    // `xmlns` declarations are also allowed at block level.
+    // The `ex` prefix is also bound to `"http://example.com"`.
```
Comment (praneesha) [inline examples/xmlns-declarations/xmlns_declarations.bal:12]:
```suggestion
    // The `ex` prefix is also bound to `http://example.com`.
```

File: examples/xmlns-declarations/xmlns_declarations.bal [hunk 21]
```diff
@@ -1,17 +1,25 @@
...
+    xmlns "http://example.com" as ex;
+
+    // The prefixes can also be used in XML navigation. 
+    boolean b = x === x.<ex:doc>;
+
+    // Since both `eg` and `ex` bind to the same namespace URL the following evaluates to `true`.
```
Comment (praneesha) [inline examples/xmlns-declarations/xmlns_declarations.bal:18]:
```suggestion
    // Since both `eg` and `ex` bind to the same namespace URL, the following evaluates to `true`.
```

File: examples/xmlns-declarations/xmlns_declarations.md [hunk 22]
```diff
@@ -1,9 +1,10 @@
 # XMLNS declarations
 
-The `xmlns` declarations are like import declarations, but bind the prefix to a namespace URL rather than
-a module. The `xmlns` declarations in the Ballerina module provide namespace context for parsing `xml`
-templates. The Qualified names in Ballerina modules are expanded into `strings` using the `xmlns`
-declarations in the module. The `xmlns` declarations are also allowed at block level.
+`xmlns` declarations are like import declarations, but bind the prefix to a namespace URL rather than a module.
```
Comment (praneesha) [inline examples/xmlns-declarations/xmlns_declarations.md:3]:
```suggestion
`xmlns` declarations are like import declarations but they bind the prefix to a namespace URL rather than a module. 
```

File: examples/xml-subtyping/xml_subtyping.md (hunk 8 above)
Comment (KRVPerera) [reply examples/xml-subtyping/xml_subtyping.md:7]: I suggested 
```suggestion
An `xml` value belongs to the type `xml<T>` if each member belongs to `T`. 
```

File: examples/xml-subtyping/xml_subtyping.bal (hunk 5 above)
Comment (KRVPerera) [reply examples/xml-subtyping/xml_subtyping.bal:31]: There is an extra space here. And do we need `type` before `T` too?
I also suggested changing to 
```suggestion
    // An `xml` value belongs to the type `xml<T>` if each member belongs to `T`.
```

File: examples/xml-operations/xml_operations.bal (hunk 6 above)
[the remark this answers is not in the document]
Comment (KRVPerera) [reply examples/xml-operations/xml_operations.bal:39]: Resolving as @praneesha also suggest the same

File: examples/xml-subtyping/xml_subtyping.bal (hunk 5 above)
Comment (praneesha) [reply examples/xml-subtyping/xml_subtyping.bal:31]: Yes. That's the convention we follow.

File: examples/xml-operations/xml_operations.bal (hunk 2 above)
Comment (KRVPerera) [reply examples/xml-operations/xml_operations.bal:58]: Resolving this

File: examples/xml-namespaces/xml_namespaces.md (hunk 11 above)
Comment (MaryamZi) [reply examples/xml-namespaces/xml_namespaces.md:3]: Will change to

```suggestion
The `ns:x` qualified name in XML is expanded into `{url}x` in which `url` is the namespace name bound to `ns`. The XML namespace declarations are kept as attributes using the standard binding of [xmlns](http://www.w3.org/2000/xmlns/).
```

File: examples/xml-subtyping/xml_subtyping.bal (hunk 4 above)
Comment (MaryamZi) [reply examples/xml-subtyping/xml_subtyping.bal:26]: How about adding
```suggestion
    // Concatenating multiple items results in a sequence of items (`xml<T>`).
    // The concatenation below will result in a sequence of XML elements (`xml<xml:Element>`).
```

File: examples/xml-subtyping/xml_subtyping.bal (hunk 5 above)
Comment (MaryamZi) [reply examples/xml-subtyping/xml_subtyping.bal:31]: IMO, when we just say "if each member" it's not clear what it is a member of.

Comment (MaryamZi) [reply examples/xml-subtyping/xml_subtyping.bal:31]: Will change to

```suggestion
    // An `xml` value belongs to the `xml<T>` type if each of its members belongs to `T`.
```

File: examples/xmlns-declarations/xmlns_declarations.bal (hunk 20 above)
Comment (MaryamZi) [reply examples/xmlns-declarations/xmlns_declarations.bal:12]: It's kinda implied that it is the string.

File: examples/xml-namespaces/xml_namespaces.bal (hunk 9 above)
Comment (praneesha) [reply examples/xml-namespaces/xml_namespaces.bal:4]: +1 to keep unchanged.

File: examples/xml-namespaces/xml_namespaces.bal (hunk 10 above)
Comment (praneesha) [reply examples/xml-namespaces/xml_namespaces.bal:5]: +1 to keep unchanged.

**Review by gimantha — APPROVED** [review]
LGTM

**Review by praneesha — APPROVED** [review]
